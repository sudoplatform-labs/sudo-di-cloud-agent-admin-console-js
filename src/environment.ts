export const uiConfig = '/uiconfig.json';
export const acapyConfig = '/acapy.json';

export interface EnvironmentInfo {
  acapyAdminUri: string; // REST API For Cloud Agent
  acapyAdminKey: string; // REST API authorisation key
  featureFlags: string[]; // Enabled/Disbled features for this customer
}
