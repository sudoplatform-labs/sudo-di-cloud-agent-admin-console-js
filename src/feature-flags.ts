/**
 * Feature flags are set during ACC deployment via runtime deployer variables.
 * E.g. RD_FF_EXAMPLE_FEATURE => FeatureFlag.EXAMPLE_FEATURE
 */
export enum FeatureFlag {
  EXAMPLE_FEATURE,
}

export function getFeatureFlags(featureFlagStrings: string[]): FeatureFlag[] {
  return featureFlagStrings
    .map((ffs) => {
      return FeatureFlag[ffs as keyof typeof FeatureFlag];
    })
    .filter((flag) => flag !== undefined)
    .sort();
}
