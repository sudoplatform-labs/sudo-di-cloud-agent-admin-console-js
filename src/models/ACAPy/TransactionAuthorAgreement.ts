// We use snake_case in the models modules for protocol related
// fields because the Aries RFCs
// define the protocol using that schema.  Originally I attempted to
// convert to camel case here to isolate that peculiarity, however that
// resulted in confusion when comparing to on-the-wire data. In addition
// because of an issue with OpenAPI code generation of typescript-fetch,
// if we don't use 'orginal' for the names output the generated
// api doesn't match the json objects returned in responses.

/*
 * This module contains all routines and definitions required to
 * interface with the Cloud Agent service for activities related to
 * Transaction Author Agreement processing.
 */
import { CloudAgentAPI } from '../../containers/App/AppContext';
import { reportCloudAgentError } from '../../utils/errorlog';
import { TAAInfo, TAAResult } from '@sudoplatform-labs/sudo-di-cloud-agent';

export type TAAAcceptParams = {
  // The way the TAA was displayed and acknowledged.
  // Must be one of the key values returned in the aml_record.aml[]
  mechanism: string;
  // The EXACT string displayed to the user and accepted.
  // This will be used in signing and must match the ledgers TAA
  text: string;
  // The taa_record.version value from the fetchLedgerTaa() result.
  version?: string;
};

export async function fetchLedgerTaa(agent: CloudAgentAPI): Promise<TAAInfo> {
  try {
    const ledgerTaa: TAAResult = await agent.ledger.ledgerTaaGet();

    return ledgerTaa.result ?? {};
  } catch (error) {
    throw await reportCloudAgentError(
      `Current Transaction Author Agreement NOT RETURNED by Ledger`,
      error,
    );
  }
}

export async function acceptLedgerTaa(
  agent: CloudAgentAPI,
  params: TAAAcceptParams,
): Promise<void> {
  try {
    await agent.ledger.ledgerTaaAcceptPost({
      body: {
        mechanism: params.mechanism,
        text: params.text,
        version: params.version,
      },
    });
  } catch (error) {
    throw await reportCloudAgentError(
      `Transaction Author Agreement NOT ACCEPTED by Ledger`,
      error,
    );
  }
}
