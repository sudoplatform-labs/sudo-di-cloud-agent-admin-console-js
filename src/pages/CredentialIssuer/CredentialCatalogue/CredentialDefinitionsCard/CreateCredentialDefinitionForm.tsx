import React, { useState } from 'react';
import {
  Form,
  Input,
  message,
  Alert,
  Button,
  Switch,
  InputNumber,
  Popover,
} from 'antd';
import { Store } from 'rc-field-form/lib/interface';
import { useAsyncFn } from 'react-use';
import styled from 'styled-components';
import { CloudAgentAPI } from '../../../../containers/App';
import {
  createCredentialDefinition,
  CredentialDefinitionCreateParams,
} from '../../../../models/ACAPy/CredentialDefinitions';
import { IndySchemaIdEntryItem } from '../../../../components/Form/IndySchemaIdEntryItem';
import { modalTAAAcceptance } from '../../../../components/Form';
import { HStack } from '../../../../components/layout-stacks';
import { InfoCircleOutlined } from '@ant-design/icons';
import { theme } from '../../../../theme';
import { RuleObject } from 'antd/lib/form';
import { StoreValue } from 'antd/lib/form/interface';

const FooterContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  margin: 0 -24px -8px;
  padding: 18px 24px 0;
`;

const StyledButton = styled(Button)`
  min-width: 112px;
`;

const SubmitButton = styled(StyledButton)`
  margin-left: 12px;
`;

/**
 * Stylised hover information icon to explain the revocable
 * credentials option and limits imposed when emabled.
 */
const StyledInfoCircleOutlined = styled(InfoCircleOutlined)`
  color: ${theme.colors.sudoBlue};
`;

const RevocableSwitchIconPopover: React.FC = () => {
  const content = (
    <p>
      Revocable credential definitions support the cancelation of an
      <br /> issued credential at a later date. Holders of revocable
      <br /> credentials can create proof presentations which verify a
      <br /> credential was valid at a point in time. Revocable
      <br /> credential definitions require a Revocable Registry definition(s)
      <br /> to be written to the ledger as well as a
      <br /> publicly accessible tails file(s).
      <br />
      <br /> The &quot;Registry Size&quot; field is only relevent for revocable
      <br /> credential definitions. It defines how many credentials
      <br /> can be issued before a new Revocable Registry definition must
      <br /> be written to the ledger (i.e. incurring a cost). It also
      <br /> defines how large the tails file will be that holders must
      <br /> download when creating a proof presentation to a verifier
      <br /> (i.e. a 1000 entry tails file is approximately 250KB).
    </p>
  );
  return (
    <Popover
      id="RevocableSwitch__popover-dialog"
      title="Revocable Credential Definitions"
      trigger="hover"
      content={content}>
      <StyledInfoCircleOutlined id="RevocableSwitch__popover-icon" />
    </Popover>
  );
};

interface Props {
  onSuccess: () => void;
  onCancel: () => void;
  cloudAgentAPIs: CloudAgentAPI;
}

export const CreateCredentialDefinitionForm: React.FC<Props> = (props) => {
  const { onCancel, onSuccess, cloudAgentAPIs } = props;
  const [form] = Form.useForm();
  const [enableSize, setEnableSize] = useState(false);

  const [submitState, handleSubmit] = useAsyncFn(
    // On initial entry we don't want to do anything
    async (values?: Store) => {
      if (!values) {
        return undefined;
      }

      const params: CredentialDefinitionCreateParams = {
        tag: values.tag,
        schema: values.schemaId,
        revocable: values.revocable,
        size: values.size,
      };

      try {
        await modalTAAAcceptance(cloudAgentAPIs, async () => {
          await createCredentialDefinition(cloudAgentAPIs, params);
          message.success('Credential Definition Created!');
          form.resetFields();
          setEnableSize(false);
          onSuccess();
        });
      } catch {
        message.error('Unable to write to public ledger, please try again');
      }
    },
    [cloudAgentAPIs, form, onSuccess],
  );

  return (
    <Form
      id="CreateCredentialDefinitionForm"
      form={form}
      initialValues={{ size: 10000 }}
      onFinish={handleSubmit}
      layout="vertical">
      <Form.Item
        name="tag"
        label="Credential Name"
        rules={[
          {
            required: true,
            validator: (rule: RuleObject, value: StoreValue): Promise<void> => {
              const validTagRegex = /^[a-zA-Z0-9_-]+$/;
              // Vaidate that the pattern matches a valid schema Id before attempting
              // to fetch it to avoid constant fetch attempts on every character entry.
              if (value && validTagRegex.test(value)) {
                return Promise.resolve(value);
              } else {
                return Promise.reject(
                  'Credential names can only contain the following characters a-zA-Z0-9_-',
                );
              }
            },
          },
        ]}>
        <Input type="text" placeholder="e.g. ACME_University_Transcript  " />
      </Form.Item>
      <IndySchemaIdEntryItem name="schemaId" />
      <HStack spacing={40}>
        <Form.Item
          name="revocable"
          label={
            <span>
              Revocable <RevocableSwitchIconPopover />
            </span>
          }
          valuePropName="checked">
          <Switch onChange={(checked) => setEnableSize(checked)} />
        </Form.Item>
        <Form.Item name="size" label="Registry Size">
          <InputNumber
            min={10}
            max={100000}
            step={1000}
            disabled={!enableSize}
          />
        </Form.Item>
      </HStack>
      {submitState.error && (
        <Alert
          type="error"
          message="There was an error creating your Verifiable Credential Definition. Please try again."
          description={submitState.error.message}
        />
      )}
      <FooterContainer>
        <StyledButton
          id="CreateCredentialDefinitionForm__cancel-btn"
          onClick={() => {
            submitState.error = undefined;
            form.resetFields();
            setEnableSize(false);
            onCancel();
          }}
          type="default"
          shape="round">
          Cancel
        </StyledButton>
        <SubmitButton
          id="CreateCredentialDefinitionForm__submit-btn"
          type="primary"
          htmlType="submit"
          loading={submitState.loading}
          shape="round">
          Create
        </SubmitButton>
      </FooterContainer>
    </Form>
  );
};
