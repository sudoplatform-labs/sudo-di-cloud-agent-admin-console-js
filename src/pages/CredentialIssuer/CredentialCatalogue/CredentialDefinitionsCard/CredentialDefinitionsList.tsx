import Table, { ColumnProps, TableProps } from 'antd/lib/table';
import React, { useState } from 'react';
import styled from 'styled-components';
import {
  consoleTableMixin,
  SearchState,
  getColumnSearchProps,
} from '../../../../components/table';
import { PlusSquareOutlined, MinusSquareOutlined } from '@ant-design/icons';
import { VStack } from '../../../../components/layout-stacks';
import { Heading } from '../../../../components/charts';
import { CredentialDefinition } from '@sudoplatform-labs/sudo-di-cloud-agent';

const CredentialDefinitionsInfoTable = Table as React.FC<
  TableProps<CredentialDefinition>
>;

const StyledConsoleTable = styled(CredentialDefinitionsInfoTable)`
  ${consoleTableMixin}
  & tr > td,
  tr > th {
    padding: 16px 24px;
  }
  ul.ant-pagination.ant-table-pagination {
    margin-right: 24px;
  }
`;

interface Props {
  dataSource: CredentialDefinition[];
  loading?: boolean;
}

export const CredentialDefinitionsList: React.FC<Props> = (props) => {
  const [searchState, setSearchState] = useState<SearchState>({
    searchText: '',
    searchedColumn: '',
  });

  const makeColumns = (): ColumnProps<CredentialDefinition>[] => {
    return [
      {
        title: 'Name',
        dataIndex: 'tag',
        ...getColumnSearchProps('tag', searchState, setSearchState),
      },
      {
        title: 'Revocable',
        width: '10%',
        render(_, credentialDefinition) {
          if (credentialDefinition.value?.revocation !== undefined) {
            return 'Yes';
          } else {
            return 'No';
          }
        },
        sorter: (a, b) =>
          a.value?.revocation
            ? b.value?.revocation
              ? 0
              : 1
            : b.value?.revocation
            ? -1
            : 0,
        sortDirections: ['descend', 'ascend', 'descend'],
        defaultSortOrder: 'ascend',
      },
      {
        title: 'Signature Type',
        dataIndex: 'type',
        ...getColumnSearchProps('type', searchState, setSearchState),
      },
      {
        title: 'Schema Sequence',
        dataIndex: 'schemaId',
        ...getColumnSearchProps('schemaId', searchState, setSearchState),
      },
    ];
  };

  const columns = makeColumns();

  return (
    <StyledConsoleTable
      id="CredentialDefinitionsList"
      loading={props.loading}
      columns={columns}
      dataSource={props.dataSource}
      rowKey={(record) => record.id ?? 'Missing Id!'}
      expandable={{
        expandedRowRender: (record) => (
          <VStack>
            <Heading> Credential Definition Identifier </Heading>
            <p> {record.id} </p>
            <Heading> Ledger Node Protocol Version </Heading>
            <p> {record.ver} </p>
            <Heading> Values </Heading>
            <pre>{JSON.stringify(record.value, undefined, 2)}</pre>
          </VStack>
        ),
        expandIcon: ({ expanded, onExpand, record }) => {
          if (expanded) {
            return <MinusSquareOutlined onClick={(e) => onExpand(record, e)} />;
          } else
            return <PlusSquareOutlined onClick={(e) => onExpand(record, e)} />;
        },
      }}
    />
  );
};
