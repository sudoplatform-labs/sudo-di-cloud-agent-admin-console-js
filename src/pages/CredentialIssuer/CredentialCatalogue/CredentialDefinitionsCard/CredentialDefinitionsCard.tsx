import { InfoCircleOutlined } from '@ant-design/icons';
import { Button, Popover, Modal } from 'antd';
import React, { useCallback, useContext, useEffect, useState } from 'react';
import { useAsyncFn } from 'react-use';
import styled from 'styled-components';
import { ConsoleCard } from '../../../../components/ConsoleCard';
import { CardsCol, CardsRow } from '../../../../components/NavLayout';
import { theme } from '../../../../theme';
import { AppContext } from '../../../../containers/App';
import { CreateCredentialDefinitionForm } from './CreateCredentialDefinitionForm';
import { CredentialDefinitionsList } from './CredentialDefinitionsList';
import { fetchAllAgentCredentialDefinitionDetails } from '../../../../models/ACAPy/CredentialDefinitions';
import { CredentialDefinition } from '@sudoplatform-labs/sudo-di-cloud-agent';

type ModalState = 'open' | 'closed';

/**
 * Stylised hover information icon to explain the purpose of
 * Credential Definitions in the Decentralized Identification ecosystem.
 */
const StyledInfoCircleOutlined = styled(InfoCircleOutlined)`
  color: ${theme.colors.sudoBlue};
`;

export const CredentialDefinitionsIconPopover: React.FC = () => {
  const content = (
    <p>
      Verifiable Credential Definitions bind an Issuer DID,
      <br /> cryptographically to an existing Schema Definition.
      <br /> Only the creator of a Verifiable Credential Definition
      <br /> can successfully issue Credentials for that Definition.
    </p>
  );
  return (
    <Popover
      id="CredentialDefinitionsCard__popover-dialog"
      title="Credential Definitions"
      trigger="hover"
      content={content}>
      <StyledInfoCircleOutlined id="CredentialDefinitionsCard__popover-icon" />
    </Popover>
  );
};

/**
 * The CredentialDefinitionsCard React component deals with writing
 * Verifiable Credential Definitions to the Decentralized Ledger.
 * These are Publicly Visible and bind the credential definition
 * to the DID which was used to create it.  Only the OWNER of the
 * creating DID can ever issue valid verifiable credentials using
 * the definition.
 */
export const CredentialDefinitionsCard: React.FC = () => {
  const { cloudAgentAPIs } = useContext(AppContext);
  const [modalState, setModalState] = useState<ModalState>('closed');

  const [
    { loading: infoLoading, value: credentials, error: agentFailed },
    getCredentialDefinitionsInfo,
  ] = useAsyncFn(
    async (): Promise<CredentialDefinition[]> =>
      fetchAllAgentCredentialDefinitionDetails(cloudAgentAPIs),
    [cloudAgentAPIs],
  );

  useEffect(() => {
    getCredentialDefinitionsInfo();
  }, [getCredentialDefinitionsInfo, modalState]);

  const modalCancelHandler = useCallback(() => {
    setModalState('closed');
  }, []);

  const modalSuccessHandler = useCallback(() => {
    setModalState('closed');
  }, []);

  let credentialDefinitionsData;
  if (agentFailed) {
    credentialDefinitionsData = (
      <div>Unable to connect to the Cloud Agent Service</div>
    );
  } else {
    credentialDefinitionsData = (
      <CredentialDefinitionsList
        dataSource={credentials ?? []}
        loading={infoLoading}
      />
    );
  }

  return (
    <CardsRow>
      <CardsCol span={24}>
        <ConsoleCard
          id="CredentialDefinitionsCard"
          title={
            <span>
              My Credential Definitions <CredentialDefinitionsIconPopover />
            </span>
          }
          extra={
            <Button
              id="CredentialDefinitionsCard__create-btn"
              type="primary"
              shape="round"
              onClick={() => setModalState('open')}
              disabled={infoLoading}>
              Create Credential Definition
            </Button>
          }>
          {credentialDefinitionsData}
        </ConsoleCard>
      </CardsCol>
      <Modal
        title="Create Credential Definition"
        visible={modalState === 'open'}
        onCancel={modalCancelHandler}
        footer={null}>
        <CreateCredentialDefinitionForm
          onCancel={modalCancelHandler}
          onSuccess={modalSuccessHandler}
          cloudAgentAPIs={cloudAgentAPIs}
        />
      </Modal>
    </CardsRow>
  );
};
