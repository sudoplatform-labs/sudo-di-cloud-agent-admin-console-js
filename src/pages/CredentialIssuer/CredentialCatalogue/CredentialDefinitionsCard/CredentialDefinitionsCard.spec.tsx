import React from 'react';
import { CredentialDefinitionsCard } from './CredentialDefinitionsCard';
import { acwMount } from '../../../../test/enzyme';
import * as modelCredentialDefinitions from '../../../../models/ACAPy/CredentialDefinitions';
import * as modelTAA from '../../../../models/ACAPy/TransactionAuthorAgreement';
import { CredentialDefinitionsList } from './CredentialDefinitionsList';
import {
  AppState,
  AppContext,
  CloudAgentAPI,
} from '../../../../containers/App/AppContext';
import { EnvironmentInfo } from '../../../../environment';
import * as FeatureFlags from '../../../../feature-flags';
import {
  WalletApi,
  LedgerApi,
  ConnectionApi,
  IntroductionApi,
  TrustpingApi,
  SchemaApi,
  CredentialDefinitionApi,
  IssueCredentialApi,
  RevocationApi,
  CredentialsApi,
  PresentProofApi,
  CredentialDefinition,
  TAAInfo,
} from '@sudoplatform-labs/sudo-di-cloud-agent';
import { Button, Modal } from 'antd';
import { act } from 'react-dom/test-utils';
import { changeFormInput, submitForm } from '../../../../test/form';

const mockEnvironment: EnvironmentInfo = {
  acapyAdminUri: 'localhost:3000/',
  acapyAdminKey: 'testkey',
  featureFlags: ['EXAMPLE_FEATURE'],
};

const mockContext: AppState = {
  environment: mockEnvironment,
  featureFlags: [FeatureFlags.FeatureFlag.EXAMPLE_FEATURE],
  cloudAgentAPIs: {
    wallet: new WalletApi(),
    ledger: new LedgerApi(),
    introductions: new IntroductionApi(),
    connections: new ConnectionApi(),
    ping: new TrustpingApi(),
    defineSchemas: new SchemaApi(),
    defineCredentials: new CredentialDefinitionApi(),
    issueCredentials: new IssueCredentialApi(),
    revocations: new RevocationApi(),
    credentials: new CredentialsApi(),
    proofs: new PresentProofApi(),
    httpOptionOverrides: {
      httpPostOptionOverrides: {},
    },
  },
};

const mockCredentialDefinitions: CredentialDefinition[] = [
  {
    id: 'TEST_CREDENTIAL_ID_0',
    schemaId: 'TEST_SCHEMA_ID_0',
    type: 'CL',
    value: 'TEST_VALUE_0',
    tag: 'TEST_CREDENTIAL_TAG_0',
    ver: 'TEST_LEDGER_NODE_VERSION',
  },
];

const fetchAllAgentCredentialDefinitionDetailsSpy = jest
  .spyOn(modelCredentialDefinitions, 'fetchAllAgentCredentialDefinitionDetails')
  .mockResolvedValue(mockCredentialDefinitions);

const createCredentialDefinitionSpy = jest
  .spyOn(modelCredentialDefinitions, 'createCredentialDefinition')
  .mockImplementation(
    async (
      cloudAgentAPI: CloudAgentAPI,
      credDefInfo: modelCredentialDefinitions.CredentialDefinitionCreateParams,
    ) => {
      const newCredDef: CredentialDefinition = {
        id: 'TEST_CREDENTIAL_ID_1',
        schemaId: credDefInfo.schema,
        type: 'CL',
        value: 'TEST_VALUE_1',
        tag: credDefInfo.tag,
        ver: 'TEST_LEDGER_NODE_VERSION',
      };
      mockCredentialDefinitions.push(newCredDef);
    },
  );

const mockTaa: TAAInfo = {
  taa_required: false,
};

const fetchLedgerTaaSpy = jest
  .spyOn(modelTAA, 'fetchLedgerTaa')
  .mockResolvedValue(mockTaa);

describe('CredentialDefinitionsCard', () => {
  it('should get credential list', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <CredentialDefinitionsCard />
      </AppContext.Provider>,
    );
    const credentialDefinitionsList = component.find(CredentialDefinitionsList);
    expect(fetchAllAgentCredentialDefinitionDetailsSpy).toBeCalled();
    expect(credentialDefinitionsList.prop('dataSource')).toBe(
      mockCredentialDefinitions,
    );
  });

  it('should expand a credential definition row', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <CredentialDefinitionsCard />
      </AppContext.Provider>,
    );

    const credentialDefinitionRow = component
      .find(CredentialDefinitionsList)
      .findWhere((node) => {
        return node.prop('rowKey') === mockCredentialDefinitions[0].id;
      });

    await act(async () => {
      credentialDefinitionRow.find('AntdIcon').simulate('click');
    });

    component.update();
    expect(credentialDefinitionRow).toMatchSnapshot();
    // JC*** how do you check expansion happened ???
  });

  it('should open, add credential definition modal', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <CredentialDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'CredentialDefinitionsCard__create-btn',
      );

    await act(async () => {
      addButton.simulate('click');
    });

    component.update();
    const modal = component.find(Modal);
    expect(modal.prop('visible')).toBe(true);
  });

  it('should close, add credential definition modal', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <CredentialDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'CredentialDefinitionsCard__create-btn',
      );
    await act(async () => {
      addButton.simulate('click');
    });
    component.update();
    const modal = component.find(Modal);
    expect(modal.prop('visible')).toBe(true);

    const cancelButton = modal
      .find(Button)
      .filterWhere(
        (node) =>
          node.prop('id') === 'CreateCredentialDefinitionForm__cancel-btn',
      );
    await act(async () => {
      cancelButton.simulate('click');
    });
    component.update();
    const updatedModal = component.find(Modal);
    expect(updatedModal.prop('visible')).toBe(false);
  });

  it('should add a new credential definition and display', async () => {
    const component = await acwMount(
      <AppContext.Provider value={mockContext}>
        <CredentialDefinitionsCard />
      </AppContext.Provider>,
    );
    const addButton = component
      .find(Button)
      .filterWhere(
        (node) => node.prop('id') === 'CredentialDefinitionsCard__create-btn',
      );
    await act(async () => {
      addButton.simulate('click');
    });
    component.update();
    const credentialForm = component
      .find(Modal)
      .filterWhere(
        (node) => node.prop('title') === 'Create Credential Definition',
      );
    expect(credentialForm.prop('visible')).toBe(true);

    await act(async () => {
      await changeFormInput(credentialForm, 'tag', 'TEST_CREDENTIAL_TAG_NEW');
      // Must be a valid schema syntax because it is validated by the form !
      await changeFormInput(
        credentialForm,
        'schemaId',
        'NYBvswA8j4ZfvFRhZLqnov:2:Drivers License:1.0',
      );
    });

    await submitForm(credentialForm);
    component.update();
    const updatedForm = component
      .find(Modal)
      .filterWhere(
        (node) => node.prop('title') === 'Create Credential Definition',
      );
    expect(fetchAllAgentCredentialDefinitionDetailsSpy).toBeCalledTimes(3);
    expect(fetchLedgerTaaSpy).toBeCalledTimes(1);
    expect(createCredentialDefinitionSpy).toBeCalledTimes(1);
    expect(updatedForm.prop('visible')).toBe(false);
    const credentialDefinitionsList = component.find(CredentialDefinitionsList);
    expect(credentialDefinitionsList.prop('dataSource')).toBe(
      mockCredentialDefinitions,
    );
  });
});
