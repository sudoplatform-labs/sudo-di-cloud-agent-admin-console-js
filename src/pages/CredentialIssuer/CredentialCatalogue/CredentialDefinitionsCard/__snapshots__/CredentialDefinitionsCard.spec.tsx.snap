// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialDefinitionsCard should expand a credential definition row 1`] = `
<BodyRow
  cellComponent="td"
  childrenColumnName="children"
  expandedKeys={Set {}}
  getRowKey={[Function]}
  index={0}
  key="TEST_CREDENTIAL_ID_0"
  record={
    Object {
      "id": "TEST_CREDENTIAL_ID_0",
      "schemaId": "TEST_SCHEMA_ID_0",
      "tag": "TEST_CREDENTIAL_TAG_0",
      "type": Object {},
      "value": Object {},
      "ver": "TEST_LEDGER_NODE_VERSION",
    }
  }
  recordKey="TEST_CREDENTIAL_ID_0"
  rowComponent="tr"
  rowKey="TEST_CREDENTIAL_ID_0"
>
  <tr
    className="ant-table-row ant-table-row-level-0"
    data-row-key="TEST_CREDENTIAL_ID_0"
    onClick={[Function]}
    style={Object {}}
  >
    <Cell
      className="ant-table-row-expand-icon-cell"
      component="td"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="RC_TABLE_KEY"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "id": "TEST_CREDENTIAL_ID_0",
          "schemaId": "TEST_SCHEMA_ID_0",
          "tag": "TEST_CREDENTIAL_TAG_0",
          "type": Object {},
          "value": Object {},
          "ver": "TEST_LEDGER_NODE_VERSION",
        }
      }
      render={[Function]}
    >
      <td
        className="ant-table-cell ant-table-row-expand-icon-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        <ForwardRef(PlusSquareOutlined)
          onClick={[Function]}
        >
          <AntdIcon
            icon={
              Object {
                "icon": Object {
                  "attrs": Object {
                    "focusable": "false",
                    "viewBox": "64 64 896 896",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "d": "M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z",
                      },
                      "tag": "path",
                    },
                    Object {
                      "attrs": Object {
                        "d": "M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z",
                      },
                      "tag": "path",
                    },
                  ],
                  "tag": "svg",
                },
                "name": "plus-square",
                "theme": "outlined",
              }
            }
            onClick={[Function]}
          >
            <span
              aria-label="plus-square"
              className="anticon anticon-plus-square"
              onClick={[Function]}
              role="img"
              tabIndex={-1}
            >
              <IconReact
                icon={
                  Object {
                    "icon": Object {
                      "attrs": Object {
                        "focusable": "false",
                        "viewBox": "64 64 896 896",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "d": "M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z",
                          },
                          "tag": "path",
                        },
                        Object {
                          "attrs": Object {
                            "d": "M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z",
                          },
                          "tag": "path",
                        },
                      ],
                      "tag": "svg",
                    },
                    "name": "plus-square",
                    "theme": "outlined",
                  }
                }
              >
                <svg
                  aria-hidden="true"
                  data-icon="plus-square"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  key="svg-plus-square"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M328 544h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8z"
                    key="svg-plus-square-svg-0"
                  />
                  <path
                    d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                    key="svg-plus-square-svg-1"
                  />
                </svg>
              </IconReact>
            </span>
          </AntdIcon>
        </ForwardRef(PlusSquareOutlined)>
      </td>
    </Cell>
    <Cell
      component="td"
      dataIndex="tag"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="tag"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "id": "TEST_CREDENTIAL_ID_0",
          "schemaId": "TEST_SCHEMA_ID_0",
          "tag": "TEST_CREDENTIAL_TAG_0",
          "type": Object {},
          "value": Object {},
          "ver": "TEST_LEDGER_NODE_VERSION",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        TEST_CREDENTIAL_TAG_0
      </td>
    </Cell>
    <Cell
      className=""
      component="td"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="RC_TABLE_KEY_next"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "id": "TEST_CREDENTIAL_ID_0",
          "schemaId": "TEST_SCHEMA_ID_0",
          "tag": "TEST_CREDENTIAL_TAG_0",
          "type": Object {},
          "value": Object {},
          "ver": "TEST_LEDGER_NODE_VERSION",
        }
      }
      render={[Function]}
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        No
      </td>
    </Cell>
    <Cell
      component="td"
      dataIndex="type"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="type"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "id": "TEST_CREDENTIAL_ID_0",
          "schemaId": "TEST_SCHEMA_ID_0",
          "tag": "TEST_CREDENTIAL_TAG_0",
          "type": Object {},
          "value": Object {},
          "ver": "TEST_LEDGER_NODE_VERSION",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      />
    </Cell>
    <Cell
      component="td"
      dataIndex="schemaId"
      firstFixLeft={false}
      firstFixRight={false}
      index={0}
      key="schemaId"
      lastFixLeft={false}
      lastFixRight={false}
      prefixCls="ant-table"
      record={
        Object {
          "id": "TEST_CREDENTIAL_ID_0",
          "schemaId": "TEST_SCHEMA_ID_0",
          "tag": "TEST_CREDENTIAL_TAG_0",
          "type": Object {},
          "value": Object {},
          "ver": "TEST_LEDGER_NODE_VERSION",
        }
      }
    >
      <td
        className="ant-table-cell"
        colSpan={null}
        rowSpan={null}
        style={Object {}}
      >
        TEST_SCHEMA_ID_0
      </td>
    </Cell>
  </tr>
</BodyRow>
`;
